File: ruby/lexer.rb
class Lexer
  OPERATORS = %w(+ * - /)
  OPERATOR_REGEX = Regexp.union(*OPERATORS)

  def initialize(text)
    @text = text
    @tokens = []
  end

  attr_reader :tokens

  def tokenize
    i = 0
    while i < @text.size
      chunk = @text[i..-1]
      if number = chunk[/\A([\d\.]+)/, 1]
        @tokens << [:NUMBER, number]
        i += number.size
      elsif negative_number = chunk[/\A(\-[\d\.]+)/, 1]
        @tokens << [:NUMBER, negative_number]
        i += negative_number.size
      elsif space = chunk[/\A(\s+)/, 1]
        # Ignore spaces
        i += space.size
      elsif operator = chunk[/\A(#{OPERATOR_REGEX})/, 1]
        @tokens << [:OPERATOR, operator]
        i += operator.size
      end
    end
  end
end

File: go/lexer/lexer.go
package lexer

import (
	"regexp"
)

type Token struct {
	Kind  string
	Value string
}

func isInArray(a string, list []string) bool {
	for _, b := range list {
		if b == a {
			return true
		}
	}
	return false
}

func isOperator(op string) bool {
	operators := []string{"+", "*", "-", "/", "^"}
	return isInArray(op, operators)
}

func Lex(expression string) []Token {
	spaces := regexp.MustCompile("\\s+")
	tokens := make([]Token, 0)
	split := spaces.Split(expression, -1)

	for _, chunk := range split {
		matchedNumber, _ := regexp.MatchString("\\A[0-9\\.]+", chunk)
		if matchedNumber {
			token := Token{
				Kind:  "NUMBER",
				Value: chunk,
			}
			tokens = append(tokens, token)
		} else if isOperator(chunk) {
			token := Token{
				Kind:  "OPERATOR",
				Value: chunk,
			}
			tokens = append(tokens, token)
		}
	}

	return tokens
}
